File: clojure/code-advent-2018/src/code_advent_2018/day_4/challenge.clj
(ns code-advent-2018.day-4.challenge
  (:require [java-time]
            [clojure.string :refer [split replace starts-with? split-lines]]))

(defn parse-line
  [line]
  (let [parts (split line #"\] ")
        time-string (replace (first parts) #"\[" "")
        time (java-time/local-date-time "yyyy-MM-dd HH:mm" time-string)
        action-string (second parts)
        action (cond
                 (starts-with? action-string "wakes") "wakes"
                 (starts-with? action-string "falls") "asleep"
                 :else "shift")
        guard-id (if (= action "shift") (Integer. (replace (second (split action-string #" ")) #"#" "")) -1)]
    {:time time :action action :guard-id guard-id}))

(defn challenge1
  [filename]
  (let [file-string (slurp (str "src/code_advent_2018/day_4/" filename))
        lines (map parse-line (split-lines file-string))
        sorted-lines (sort-by :time lines)
        guard-sleep-times (reduce (fn
                                    [result line]
                                    (if-let [shift? (= (:action line) "shift")]
                                      (assoc (assoc result :current-guard (:guard-id line)) :previous-time (:time line))
                                      (let [sleep? (= (:action line) "asleep")
                                            minutes-path [(:current-guard result) (if sleep? :awake :asleep)]
                                            minutes (or (get-in result minutes-path) [])
                                            new-minutes (range (.getMinute (:previous-time result)) (.getMinute (:time line)))]
                                        (assoc (assoc-in result minutes-path (apply conj minutes new-minutes)) :previous-time (:time line)))))
                                  {:current-guard -1 :previous-time -1}
                                  sorted-lines)
        most-asleep-guard-id (key (apply max-key #(count (:asleep (second %))) guard-sleep-times))
        most-asleep-minute-of-most-asleep-guard (first (last (sort-by second (frequencies (get-in guard-sleep-times [most-asleep-guard-id :asleep])))))]
    (* most-asleep-guard-id most-asleep-minute-of-most-asleep-guard)))

(defn challenge2
  [filename]
  (let [file-string (slurp (str "src/code_advent_2018/day_4/" filename))
        lines (map parse-line (split-lines file-string))
        sorted-lines (sort-by :time lines)
        guard-sleep-times (dissoc (reduce (fn
                                            [result line]
                                            (if-let [shift? (= (:action line) "shift")]
                                              (assoc (assoc result :current-guard (:guard-id line)) :previous-time (:time line))
                                              (let [sleep? (= (:action line) "asleep")
                                                    minutes-path [(:current-guard result) (if sleep? :awake :asleep)]
                                                    minutes (or (get-in result minutes-path) [])
                                                    new-minutes (range (.getMinute (:previous-time result)) (.getMinute (:time line)))]
                                                (assoc (assoc-in result minutes-path (apply conj minutes new-minutes)) :previous-time (:time line)))))
                                          {:current-guard -1 :previous-time -1}
                                          sorted-lines)
                                  :current-guard
                                  :previous-time)
        guards-most-slept-mins (map #(conj (last (sort-by second (frequencies (:asleep (second %))))) (first %)) guard-sleep-times)
        guard-most-slept-mins (last (sort-by second guards-most-slept-mins))]
    (* (first guard-most-slept-mins) (nth guard-most-slept-mins 2))))

File: 2018/python/Day4/challenge.py
from datetime import datetime

def parseLine(line):
    """
    parses a line.

    >>> parseLine("[1518-04-21 00:57] wakes up")
    {'timeString': '1518-04-21 00:57', 'time': datetime.datetime(1518, 4, 21, 0, 57), 'action': 'wakes', 'guard-id': -1}
    >>> parseLine("[1518-09-03 00:12] falls asleep")
    {'timeString': '1518-09-03 00:12', 'time': datetime.datetime(1518, 9, 3, 0, 12), 'action': 'asleep', 'guard-id': -1}
    >>> parseLine("[1518-04-21 00:04] Guard #3331 begins shift")
    {'timeString': '1518-04-21 00:04', 'time': datetime.datetime(1518, 4, 21, 0, 4), 'action': 'shift', 'guard-id': 3331}
    """
    parts = line.split("] ")
    timeString = parts[0].replace("[", "")
    time = datetime.strptime(timeString, "%Y-%m-%d %H:%M")
    actionString = parts[1]
    if actionString.startswith("wakes"):
        action = "wakes"
    elif actionString.startswith("falls"):
        action = "asleep"
    else:
        action = "shift"
    guardId = -1 if action != "shift" else int(actionString.split(" ")[1].replace("#", ""))
    return {"timeString": timeString, "time": time, "action": action, "guard-id": guardId}

def challenge1(filename):
    """
    challenge1

    >>> challenge1("input.txt")
    36898
    """
    lines = []
    with open(filename) as f:
        lines = [parseLine(line) for line in f]

    lines.sort(key=(lambda item: item["time"]))

    values = {"current-guard": -1, "previous-time": -1}
    for line in lines:
        currentTime = int(line["timeString"].split(":")[1])
        if line["action"] == "shift":
            values["current-guard"] = line["guard-id"]
        else:
            if values["current-guard"] not in values:
                values[values["current-guard"]] = {"awake": [], "asleep": []}

            values[values["current-guard"]]["awake" if line["action"] == "asleep" else "asleep"].extend(range(values["previous-time"], currentTime))

        values["previous-time"] = currentTime
    del values["current-guard"]
    del values["previous-time"]

    mostAsleepGuardId = max(values.keys(), key=(lambda key: len(values[key]["asleep"])))
    guardAsleepMinutes = values[mostAsleepGuardId]["asleep"]
    mostAsleepMinute = max(set(guardAsleepMinutes), key=guardAsleepMinutes.count)
    return mostAsleepGuardId * mostAsleepMinute

def challenge2(filename):
    """
    challenge2

    >>> challenge2("input.txt")
    80711
    """
    lines = []
    with open(filename) as f:
        lines = [parseLine(line) for line in f]

    lines.sort(key=(lambda item: item["time"]))

    values = {"current-guard": -1, "previous-time": -1}
    for line in lines:
        currentTime = int(line["timeString"].split(":")[1])
        if line["action"] == "shift":
            values["current-guard"] = line["guard-id"]
        else:
            if values["current-guard"] not in values:
                values[values["current-guard"]] = {"awake": [], "asleep": []}

            values[values["current-guard"]]["awake" if line["action"] == "asleep" else "asleep"].extend(range(values["previous-time"], currentTime))

        values["previous-time"] = currentTime
    del values["current-guard"]
    del values["previous-time"]

    mostAsleepGuardId = max(values.keys(), key=(lambda key: max(set(values[key]["asleep"]), key=values[key]["asleep"].count)))
    guardAsleepMinutes = values[mostAsleepGuardId]["asleep"]
    mostAsleepMinute = max(set(guardAsleepMinutes), key=guardAsleepMinutes.count)
    return mostAsleepGuardId * mostAsleepMinute

if __name__ == "__main__":
    import doctest
    doctest.testmod()